examples/medplum-demo-bots/src/deduplication/merge-bot.ts

```diff
@@ -0,0 +1,151 @@
+  (reports as ResourceWithSubject[]).map(async (report) => {
+    // Update each found resource's subject to the target patient.
+    report.subject = createReference(target);
+    await medplum.updateResource(report);
+  });
+}
+
+async function addToDoNotMatchList(
+  medplum: MedplumClient,
+  patientList: Reference<Patient>,
+  patientAdded: Reference<Patient>
+): Promise<void> {
+  const lists = await medplum.searchResources('List', { subject: patientList, code: 'doNotMatch' });
+  if (lists.length === 0) {
+    console.warn('No doNotMatch list found for patient: ' + patientList);
+  }
+  lists.forEach(async (list) => {
+    const entries = list.entry;
+    entries?.push({ item: patientAdded });
+    await medplum.updateResource({ ...list, entry: entries });
+  });
+}
+
+/**
+ * Handler function to process incoming BotEvent containing a QuestionnaireResponse for potential patient record merges.
+ */
+export async function handler(medplum: MedplumClient, event: BotEvent<QuestionnaireResponse>): Promise<any> {
+  // Extract answers from the QuestionnaireResponse.
+  const responses = getQuestionnaireAnswers(event.input);
+  // Get the reference to the RiskAssessment from the answers.
+  const riskAssessmentReference = responses['assessment'] as QuestionnaireResponseItemAnswer;
+  // If there's no valid RiskAssessment reference in the response, throw an error.
+  if (!riskAssessmentReference.valueReference) {
+    throw new Error('Invalid input. Expected RiskAssessment reference');
+  }
+  const riskAssessment = (await medplum.readReference(riskAssessmentReference.valueReference)) as RiskAssessment;
+  const targetReference = riskAssessment.basis?.[0] as Reference<Patient>;
+  const srcReference = riskAssessment.subject as Reference<Patient>;
+  if (!targetReference || !srcReference) {
+    throw new Error(`Undefined references target: ${targetReference} src: ${srcReference}`);
+  }
+
+  // If merge is disabled based on the questionnaire's answer, terminate the handler early.
+  const mergeCheck = responses['disableMerge']?.valueBoolean;
+  if (!!mergeCheck) {
+    await addToDoNotMatchList(medplum, srcReference, targetReference);
+    await addToDoNotMatchList(medplum, targetReference, srcReference);
+    return true;
+  }
+
+  const patientTarget = await medplum.readReference(targetReference);
+  const patientSource = await medplum.readReference(srcReference);
+
+  const patients = linkPatientRecords(patientSource as Patient, patientTarget as Patient);
+
+  const mergedPatients = mergePatientRecords(patients.src, patients.target);
+  const deleteSource = responses['deleteSource']?.valueBoolean;
+
+  await updateClinicalReferences(medplum, mergedPatients.src, mergedPatients.target, 'ServiceRequest');
+  if (deleteSource === true) {
+    await medplum.deleteResource('Patient', mergedPatients.src.id as string);
+  } else {
+    await medplum.updateResource<Patient>(mergedPatients.src);
```

Comment:
comment needed to remind readers _why_ we need to make this call to `updateResource`